### PR TITLE
Allow Excluding Starting Categories

### PR DESCRIPTION
**Pull request by donn-is-me**
Adds checkboxes to the category tracker that disables the visualization of all non-terminal starting categories, i.e., all of them except Any%.

Mainly because Hectique asked for it and I was bored.

Ran the unit tests and tested it for a while. Seems fine.

Screenshot from 2022-01-09 08-00-48

### CODE REVIEW

**Comment by donn-is-me** [comment]
Ah. Linting. Sorry. I'll reformat later today.

**Comment by gmjosack** [comment]
There is also a large change in PR right now to how we do config on the trackers to make config persistent that this is likely incompatible with.

**Comment by mauvealerts** [comment]
Thanks for sending this. It's a lot of flexibility. I'm going to mull this over a bit, mostly about:

* Interaction with making `RunLabel` support score vs speed
* Storing these settings as we add more labels (#218)
* UI structure if we add toggles for other modifiers, or the termini

For the failing checks, you just need to run `black src/` to fix the formatting. You might be able to make that run automatically (e.g. "Format on Save" in VSCode).

I was going to suggest adding tests to `src\tests\ui\trackers\runstate_test.py`, but apparently that doesn't even exercise `hide_early` 😅

**Comment by gmjosack** [comment]
Running `black` won't be sufficient for the current lint errors. passing `[]` as a default value is a footgun in python and that's what is failing. the pattern is to do something like

```py
def foo(x=None):
    if x is None:
        x = []
```

**Comment by donn-is-me** [comment]
Both addressed- I didn't know about default values being mutable in Python. That's terrifying.

**Comment by donn-is-me** [comment]
Rebased to conform with #218's changes.

**Comment by mauvealerts** [comment]
I think this is going to end up being more complicated change, for reasons I'll explain below. The result will probably be a rewrite. If you want to work on this, I'm happy to talk through the details. If you don't, I was planning to implement something like this roughly now-ish anyway.

Regardless of your choice, I appreciate you sending this. Rewriting it might have been avoided with some discussion, but I've noticed that the README doesn't indicate that, or even point folks anywhere. I'll try to fix that Soon(TM), but ... folks usually chat in the `#s2-modding-tooldevs` channel on the [Spelunky community Discord](https://discord.gg/spelunky-community).

As for changes... I'm also trying to keep the UI approachable. After a small discussion in `#s2-speedrun`, I think we only want toggles for a few labels: ones that are both unpopular and appear by default. The image below is my rough UI idea ("Run Type" selection is a related-ish thing I want to add soon).

Cat Tracker Options


I've been working on saving configs for a while, since it's already annoying to tweak the existing settings each time. Unfortunately, the plumbing is still rough. The `RunState` and `RunLabel` should probably take the (new!) `CategoryTrackerConfig` instead of passing around the separate settings. Another complication is that the existing `Label` enum isn't suitable for storage: it doesn't serialize nicely, and would probably cause a circular dependency. Considering that we only want some labels, it seems best to create a config-specific `Enum` with the few values we want, and map that to the existing `Label`.

To summarize, I think the needed changes are:

* Only have toggles for "No%", "No Gold", and "Pacifist"
* Pass `CategoryTrackerConfig` instead of separate bits of data
* Add enum for config, and stuff to map from config enum to existing `Label` enum

File: src/modlunky2/config.py [hunk 1]
```diff
@@ -140,6 +140,7 @@ def clone(self: T) -> T:
 @dataclass
 class CategoryTrackerConfig(CommonTrackerConfig):
     always_show_modifiers: bool = skip_default_field(default=False)
+    excluded_categories: Optional[List[str]] = skip_default_field(default=None)
```
Comment (mauvealerts) [inline src/modlunky2/config.py:143]: This can be a `Set[MyEnum]`, provided `MyEnum` can be serialized. An `Enum` with `str` values should work well.

I think `FrozenSet` would also be fine. It's not necessary since we clone this object before sending it to the tracker's thread (where `RunState` and `RunLabel` do stuff). `Set` might be easier to work with, tho?

File: src/modlunky2/ui/trackers/category.py [hunk 2]
```diff
@@ -33,19 +34,24 @@ def __init__(self, parent, modlunky_config: Config, *args, **kwargs):
             Image.open(ICON_PATH / "cat2.png").resize((24, 24), Image.ANTIALIAS)
         )
 
+        current_row = 0
```
Comment (mauvealerts) [inline src/modlunky2/ui/trackers/category.py:37]: I think the variable is approach is fine and used elsewhere.

As an alernative, you could make a list of widgets and then use `for row, widget in enumerate(some_widget_list): widget.grid(opts_here)`. The builtin `enumerate()` produces (index, value) pairs, i.e. you could use the index of the list item as its row number

File: src/modlunky2/ui/trackers/category.py [hunk 3]
```diff
@@ -33,19 +34,24 @@ def __init__(self, parent, modlunky_config: Config, *args, **kwargs):
...
+        self.category_button.grid(
+            row=current_row, column=0, pady=5, padx=5, sticky="nswe"
+        )
 
         self.always_show_modifiers = tk.BooleanVar()
         self.always_show_modifiers.set(
             modlunky_config.trackers.category.always_show_modifiers
         )
+
         self.always_show_modifiers_checkbox = ttk.Checkbutton(
```
Comment (mauvealerts) [inline src/modlunky2/ui/trackers/category.py:55]: I think the UI part is pretty minor, but if you want to make this more like my sketch...

* Create a subclass of `ttk.LabelFrame` to hold the "Modifiers" checkboxes, and use it here (instead of directly adding them)
* Put a `ttk.Separator` between the "Always Show" checkbox and the per-label ones

**Comment by donn-is-me** [comment]
> Only have toggles for "No%", "No Gold", and "Pacifist"
> Add enum for config, and stuff to map from config enum to existing Label enum
If we're going to go that route, might it not be better to just store three booleans? Avoid the hassle of serialization and all.

> Pass CategoryTrackerConfig instead of separate bits of data
Done! I switched that over after rebasing.

**Comment by mauvealerts** [comment]
I'll try to do a more thorough review tomorrow. For now...

> > Only have toggles for "No%", "No Gold", and "Pacifist"
> > Add enum for config, and stuff to map from config enum to existing Label enum
> 
> If we're going to go that route, might it not be better to just store three booleans? Avoid the hassle of Enum serialization and all.

Three bools works just fine. There will need to be some mapping between the config-value(s) and `Label` enum values. If statements (for bools) and a dict (for Enum) work similarly

However, there are a few things that me favor Enum. First, some notable facts...
* Enum can (surprisingly) be used as an iterable
* Serialization is free-ish with pyserde

For serialization, this Just Works with pyserde:
```py
class Foo(Enum):
  BAR = "Bar"
  BAZ = "Baz"

@deserialize
class Demo:
  foos: Set[Foo] = dataclasses.field(default_factory=set)
```

The iterable part allows you to write stuff like:
```py
for v in Foo:
  // do stuff, e.g. make checkboxes

list(Foo)

ALL_FOO = set(Foo)
```
I think the iterable is pretty handy...

* for-loop to make checkboxes
* Compute what's absent, e.g. `absent = ALL_FOO - maybe_present`
* Check that every config-Enum value is present in the mapping to RunLabel-enum (in a test)

**Comment by donn-is-me** [comment]
This is what I've reached so far:

image

As you can see, I'm still storing it as a plain ol' list of label names though. I'll have to take a look at working with Enums.

**Comment by donn-is-me** [comment]
Enum'd.

File: src/modlunky2/ui/trackers/category.py [hunk 4]
```diff
@@ -78,6 +131,10 @@ def launch(self):
             config=self.modlunky_config.trackers.category,
         )
 
+    def config_update_callback(self):
```
Comment (mauvealerts) [inline src/modlunky2/ui/trackers/category.py:134]: I'd move the `save()` call into here. I can't think of a scenario where we update the thread without saving the config.

A bonus is `CategoryModifiers` could take just the `CategoryTrackerConfig`

File: src/modlunky2/ui/trackers/category.py [hunk 5]
```diff
@@ -21,11 +21,85 @@
...
+        self.always_show_modifiers = tk.BooleanVar()
+        self.always_show_modifiers.set(
+            modlunky_config.trackers.category.always_show_modifiers
+        )
+
+        widgets = []
+        widgets.append(
+            ttk.Checkbutton(
+                self,
+                text="Always Show",
```
Comment (mauvealerts) [inline src/modlunky2/ui/trackers/category.py:40]: After some more thought, "Show all modifiers before 1-3" seems clearer? Now that the per-label checkboxes are in a row, there's room to be more verbose.

When I showed the UI sketch to my partner, this text was the part that confused them the most. The existing text "Always Show Modifiers" is leftover from the days of only supporting "No Gold" and "Pacifist" modifiers.

**Review by mauvealerts — COMMENTED** [review]
Thanks for all the adjustments. I _think_ this is very close to being ready to merge.

When I initially suggested passing `CategoryTrackerConfig`, I thought it'd be a smaller refactoring. I still think it's worth doing, but I'm happy to tidy it up later.

File: src/modlunky2/ui/trackers/category.py [hunk 6]
```diff
@@ -21,11 +21,85 @@
...
+        )
+        widgets.append(ttk.Separator(self))
+
+        # Starting Category Exclusion
+
+        ## This does not have any effect on the actual tracking logic, rather,
+        ## like always_show_modifiers, it only has an effect on the actual
+        ## text returned.
+
+        ## The checkboxes imply *inclusion*, not exclusion, so the
```
[the remark this answers is not in the document]
Comment (donn-is-me) [reply src/modlunky2/ui/trackers/category.py:55]: While seemingly so, the logic is built around excluding it from visualization rather than including it in the tracking logic, so the tracking logic can be identical regardless if you're actually running No/NG/Pacifist. I understand the tracking logic is intricate, which is why I went with the approach of using exclusions and just doing a simple translation near the UI layer.

File: src/modlunky2/ui/trackers/label.py [hunk 7]
```diff
@@ -205,11 +224,21 @@ def _percent(cls, labels: Set[Label]) -> Optional[Label]:
 
         return found
 
-    def text(self, hide_early) -> str:
-        if self._cached_text is not None and self._cached_text.hide_early == hide_early:
+    def text(self, hide_early, excluded_categories=None) -> str:
```
[the remark this answers is not in the document]
Comment (donn-is-me) [reply src/modlunky2/ui/trackers/label.py:227]: Agreed on the first two points, though on the last point if we move the config all the way into label.py, we'd also have to move a lot of the logic that goes into calculating `hide_early` into label.py or use a callback of some kind. Not ideal. What I can do however that's relatively simple is at least make sure it's `CategoryTrackerConfig` all the way to runstate.get_display.

Thanks for the heads up on the tests. Was a simple change.

File: src/modlunky2/ui/trackers/category.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mauvealerts) [reply src/modlunky2/ui/trackers/category.py:57]: Makes sense

**Review by mauvealerts — APPROVED** [review]
